st2actions/st2actions/scheduler/handler.py

```diff
@@ -0,0 +1,164 @@
+from st2common.services import action as action_service
+from st2common.services import policies as policy_service
+from st2common.persistence.liveaction import LiveAction
+from st2common.persistence.execution_queue import ExecutionQueue
+from st2common.util import action_db as action_utils
+from st2common.services.coordination import LockAcquireError
+from st2common.services import coordination
+
+__all__ = [
+    'ExecutionQueueHandler',
+    'get_handler'
+]
+
+
+LOG = logging.getLogger(__name__)
+
+
+def _next_executions():
+    """
+        Sort executions by fifo and priority and get the latest, highest priority
+        item from the queue and pop it off.
+    """
+    query = {
+        "scheduled_start_timestamp__lte": date.get_datetime_utc_now(),
+        "order_by": [
+            "scheduled_start_timestamp",
+        ]
+    }
+
+    queued_executions = ExecutionQueue.query(**query)
+    if queued_executions:
+        return queued_executions
+
+    return []
+
+
+class ExecutionQueueHandler(object):
+    def __init__(self):
+        self.message_type = LiveActionDB
+        self._shutdown = False
+
+    def loop(self):
+        LOG.debug('Entering scheduler loop')
+        while self._shutdown is not True:
+            eventlet.greenthread.sleep(0.25)
+            try:
+                with coordination.lock('st2shcedulerqueue'):
+                    queued_executions = _next_executions()
+                    for execution in queued_executions:
+                        self._handle_execution(execution)
+            except LockAcquireError as e:
+                LOG.error(e)
+
+    def _handle_execution(self, execution):
+        LOG.info('Scheduling liveaction: %s', execution.liveaction)
+        try:
+            liveaction_db = action_utils.get_liveaction_by_id(execution.liveaction)
+        except StackStormDBObjectNotFoundError:
+            LOG.exception('Failed to find liveaction %s in the database.', execution.liveaction.id)
+            ExecutionQueue.delete(execution)
+            raise
+
+        liveaction_db = self._apply_pre_run(liveaction_db, execution)
+
+        if liveaction_db:
+            if not self._exit_if_not_runnable(liveaction_db, execution):
+                self._update_to_scheduled(liveaction_db, execution)
+
+    @staticmethod
+    def _apply_pre_run(liveaction_db, execution):
+        # Apply policies defined for the action.
+        liveaction_db = policy_service.apply_pre_run_policies(liveaction_db)
+
+        LOG.info("Liveaction Status Pre-Run: %s", liveaction_db.status)
+
+        if liveaction_db.status is action_constants.LIVEACTION_STATUS_POLICY_DELAYED:
+            liveaction_db = action_service.update_status(
+                liveaction_db, action_constants.LIVEACTION_STATUS_DELAYED, publish=False
+            )
+            execution.scheduled_start_timestamp = date.append_milliseconds_to_time(
+                date.get_datetime_utc_now(),
+                500
```

Comment:
I see what we you did here.

Eventually we can utilize similar approach for retry policy.